src/mercedestrenz/data.py

```diff
@@ -52,12 +55,73 @@ def listing_search(data, budget=[0, np.Inf], model = "any", sort_feature = "odom
-    >>> result.head()
-    #to be completed with exmaple output
+    >>> # search listings within a price range
+    >>> listing_search(data, budget=[2000, 20000], model = "any", sort_feature = "odometer_mi", ascending = True)
+    >>> # search listings for a specific model and below a maximum price
+    >>> listing_search(data, budget=20000, model = "glk", sort_feature = "odometer_mi", ascending = True)
     """
-    pass
 
+    # ======= Unit tests ==========
+    # check input data is a pd.dataframe
+    if type(data) != pd.DataFrame:
+        raise Exception("The input dataset is not of Pandas DataFrame format")
+
+    # check budget is either a flort/int or a list of float/int
+    if type(budget) not in [float, int]: 
+        if len(budget) == 2:
+            for elem in budget:
+                if type(elem) not in [float, int]:
+                    raise Exception("The budget range boundaries should be floats or integers")
+        else:
+            raise Exception("Please specify your maximum budget using a number, or a range of budge using a list of numeric values.")
+
+    # model= parameter input should be a string
+    if type(model) != str:
+            raise Exception("The model parameter should have a string as the input")
+    
+    # The string should correspond to a specific car model in the model column
+    if ((data['model'].eq(model)).any() or model == 'any') == False:
+        raise Exception("The specified car model does not exist in the dataframe provided")
+
+    # sort_feature should be a string
+    if type(sort_feature) != str:
+        raise Exception("The input value for sort_feature parameter should be a string")
+    
+    # sort_feature should indicate a numeric column in the provided data frame
+    numeric_cols = data.select_dtypes('number').columns.to_list()
+    if sort_feature not in numeric_cols:
+        raise Exception("The specified sort_feature should be a numeric column in the provided dataframe.")
+
+    # ascending is either true or false
+    if type(ascending) != bool:
+        raise Exception("Please specify True or False in the ascending parameter")
+    
+    # price_col string
+    if type(price_col) != str:
+        raise Exception("Please specify the price column name using a string, or use the default value.")
+
+
+    # ======= Function ==========
+    
+    # filter by budget
+    if type(budget) in [float, int]: # max budget specified
+        condition = data[price_col] <= budget
+    else:
+        condition = ((budget[0] <= data[price_col]) & (data[price_col] <= budget[1]))
+    temp_df = data.loc[condition]
+
+    # filter by model
+    if model != "any":
+        condition = data['model'] == model
+        temp_df = temp_df.loc[condition]
+
+    # sort by price & output
+    result = temp_df.sort_values(by = [sort_feature, price_col], ascending = [ascending, True])
+
+    # order output 
+    priority_order = ['url', price_col, 'model', sort_feature]
```

Comment:
Unfortunately, the `url` column was removed from the sample data. Maybe this could be replaced with the `description` column?
```suggestion
priority_order = [price_col, 'model', sort_feature, 'description']
```